messaging/src/main/java/com/flipkart/varadhi/produce/services/ProducerService.java

```diff
@@ -1,100 +1,81 @@
             Message message,
             String varadhiTopicName,
             ProduceContext context
     ) {
-        String messageId = message.getMessageId();
         try {
             String produceRegion = context.getTopicContext().getRegion();
             InternalTopic internalTopic = internalTopicCache.getProduceTopicForRegion(varadhiTopicName, produceRegion);
-
-            return produceToTopic(message, internalTopic).thenApply(result -> {
-                sendProduceMetric(messageId, result.getProduceStatus().status(), result.getProducerLatency(), context);
-                return result;
-            });
+            if (!internalTopic.getTopicState().isProduceAllowed()) {
+                return CompletableFuture.completedFuture(
+                        ProduceResult.ofNonProducingTopic(message.getMessageId(), internalTopic.getTopicState()));
+            }
+            Producer producer = producerCache.getProducer(internalTopic.getStorageTopic());
+            return produceToStorageProducer(
+                    producer, context, internalTopic.getStorageTopic().getName(), message).thenApply(result ->
+                    ProduceResult.of(message.getMessageId(), result));
         } catch (VaradhiException e) {
-            sendProduceMetric(messageId, Failed, 0, context);
             throw e;
         } catch (Exception e) {
-            sendProduceMetric(messageId, Failed, 0, context);
-            throw new ProduceException(String.format("Produce failed due to internal error."), e);
+            throw new ProduceException("Produce failed due to internal error.", e);
         }
     }
 
-    private void sendProduceMetric(
-            String messageId,
-            ProduceResult.Status status,
-            int produceLatency,
-            ProduceContext context
+
+    private CompletableFuture<AsyncResult<ProducerResult>> produceToStorageProducer(
+            Producer producer, ProduceContext context, String topic, Message message
     ) {
+        long produceStart = System.currentTimeMillis();
+        return producer.ProduceAsync(message).handle((result, throwable) -> {
+            int producerLatency = (int) (System.currentTimeMillis() - produceStart);
+            emitProducerMetric(message.getMessageId(), result != null, producerLatency, context);
+            if (throwable != null) {
+                log.debug(
+                        String.format("Produce Message(%s) to StorageTopic(%s) failed.", message.getMessageId(), topic),
+                        throwable
+                );
+            }
+            return AsyncResult.of(result, throwable);
+        });
+    }
+
+    private void emitProducerMetric(String messageId, boolean succeeded, int produceLatency, ProduceContext context) {
         try {
-            metricProvider.OnProduceEnd(messageId, status, produceLatency, context);
+            metricProvider.OnMessageProduced(succeeded, produceLatency, context);
         } catch (Exception e) {
```

Comment:
we shouldn't even catch "any" exception here.

Reply:
handle it inside metric provider or fail produce ?